# Лабораторная работа №5

**Тема:** работа с файлами в Linux средствами ассемблера

**Вариант:** 23

**Цель работы:** написать программы на FASM для обработки текстовых файлов. Имена файлов и параметры передаются через командную строку.


## Выбор задач

Вариант задаётся начальным значением генератора случайных чисел. По алгоритму из задания для `Variant = 23` выбраны задачи **5 и 2**.


In [ ]:
import random as rd
Variant = 23
rd.seed(Variant)
Numbers_of_problems = [i + 1 for i in rd.sample(range(11), 2)]
Numbers_of_problems

[5, 2]

## Задача 5. Раскачивающийся порядок символов

Для исходного файла и позиции `k` записать символы в порядке `k, k+1, k-1, k+2, k-2, …, k+m, k-m`. Позиции в файле считаются с единицы. Если позиция меньше 1 или выходит за конец файла, соответствующий символ пропускается.

**Алгоритм:** сначала прочитать символ в позиции `k`. Затем для каждого `j` от 1 до `m` прочитать позиции `k+j` и `k-j` и записать найденные символы в выходной файл. Доступ к позиции выполняется системным вызовом `lseek`, чтение и запись — вызовами `read` и `write`.

**Запуск:** `./task5 input.txt output.txt k m`

Исходный код: `task5.asm`.


In [ ]:
text = '123456789'
k, m = 5, 4
positions = [k] + [p for j in range(1, m + 1) for p in (k + j, k - j)]
''.join(text[p - 1] for p in positions if 1 <= p <= len(text))

'564738291'

## Задача 2. Каждый k-й символ

Прочитать из заданного текстового файла каждый `k`-й символ и записать его во второй файл. Позиции считаются от начала файла с единицы: выбираются символы на позициях `k, 2k, 3k, …`. Переводы строк также считаются символами файла.

**Алгоритм:** читать файл по одному байту, считать позицию, и записывать байт, когда номер позиции кратен `k`.

**Запуск:** `./task2 input.txt output.txt k`

Исходный код: `task2.asm`.


In [ ]:
text = 'abcdefghij'
k = 3
text[k - 1::k]

'cfi'

## Сборка и проверка

Сборка выполняется в Docker-среде проекта через `./run_fasm.sh`; файлы `task2.asm` и `task5.asm` собираются командой `fasm`. Проверены примеры из отчёта: для задачи 2 вход `abcdefghij`, `k = 3` даёт `cfi`; для задачи 5 вход `123456789`, `k = 5`, `m = 4` даёт `564738291`. Оба результата получены при запуске FASM-программ в контейнере.

## Вывод

Подготовлены две программы на FASM для обработки файлов. Первая извлекает каждый `k`-й символ, вторая записывает символы в порядке, расходящемся от заданной позиции влево и вправо. Параметры передаются через командную строку.
